### PR TITLE
Add `pulumi-version` option

### PR DESCRIPTION
**Pull request by lomholdt**
This PR adds the `pulumi-version` option allowing pinning the CLI version that get's installed while running the action.
If the `pulumi-version` option is not set it will default to `^3` as this was the previously hardcoded value.

### CODE REVIEW

**Comment by Moon1706** [comment]
@lomholdt Hi!
As I know, additional functionality really slooooowly appended to this Action. I try to add configMap 2 month. So, if it's critical for you, please, use [my fork](https://github.com/terraform-gitlab-modules/actions) to add new features. I hope in the future it will be an append to the official action.

**Comment by RobbieMcKinstry** [comment]
Hi @lomholdt , thank you for your contribution! As I recall, we have a similar parameter in the Azure DevOps Pipeline CI integration, so adding the Pulumi version sounds like it makes sense here too.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 - fix: update PR comments correctly when `edit-pr-comment` is true (fixes
   [#633](https://github.com/pulumi/actions/issues/633))
 
+- Add `pulumi-version` option to allow pinning the version of the CLI
```
Comment (RobbieMcKinstry) [inline CHANGELOG.md:8]:
```suggestion
- enhancement: Add `pulumi-version` option to allow pinning the version of the CLI
   [PR](https://github.com/pulumi/actions/pull/661) fixes [#437 ](https://github.com/pulumi/actions/issues/437)
```

File: README.md [hunk 2]
```diff
@@ -75,10 +75,9 @@ The action can be configured with the following arguments:
   this update
 
 - `edit-pr-comment` - (optional) Edit previous PR comment instead of posting new
-  one.  
-  **PLEASE NOTE:** that as of 3.2.0 of the Action, this now defaults to `true`.
-  This is in an effort to reduce verbosity - if you want to have a comment per
-  PR run, please ensure that you set this to `false`.
+  one. **PLEASE NOTE:** that as of 3.2.0 of the Action, this now defaults to
+  `true`. This is in an effort to reduce verbosity - if you want to have a
+  comment per PR run, please ensure that you set this to `false`.
```
Comment (RobbieMcKinstry) [inline README.md:80]: Ah, these changes seem rather unrelated to the PR at hand. AFAIK they shouldn't impact markdown rendering at all. Would you mind backing them out?

**Review by RobbieMcKinstry — CHANGES_REQUESTED** [review]

**Comment by RobbieMcKinstry** [comment]
Also, it looks like there was some discussion on https://github.com/pulumi/actions/pull/661 about generating a warning if the CLI version number is incompatible with the automation API version number.
@lomholdt Would you add a warning if the version specified is beneath `v3`?

**Comment by jonas-lomholdt** [comment]
> Hi @lomholdt , thank you for your contribution! As I recall, we have a similar parameter in the Azure DevOps Pipeline CI integration, so adding the Pulumi version sounds like it makes sense here too.
> 
> 
> 
> I'm seeing this branch has some conflicts. Would you be able to rebase? 

Sure, I will fix that. Thanks 👍🏼

**Comment by jonas-lomholdt** [comment]
> Also, it looks like there was some discussion on https://github.com/pulumi/actions/pull/661 about generating a warning if the CLI version number is incompatible with the automation API version number.
> 
> @lomholdt Would you add a warning if the version specified is beneath `v3`? Thank you very much! :D
> 
> Please re-request my review when you've made these changes. I'm going to be away from my computer for a few days for the holiday, but I can take a look when I get back. :) 

Sure, that makes sense. I will give it a shot. Thanks. Have a nice holiday 👍🏼

**Comment by RobbieMcKinstry** [comment]
Would you also remove the `dist` file updates from your PR? I believe the TypeScript files are automatically recompiled by a GitHub Action once a commit lands on `master/main`.

**Comment by lomholdt** [comment]
> Hey @jonas-lomholdt, just following up on this. Let me know if you need any assistance with the remaining action items. :)

Hey, sorry I have been taking some vacation. I fixed the conflicts and need to warn if the version is less than 3. I would probably need to parse the input string somehow to get a semver representation of it that I can compare. Did you have any specific way in mind? low-fi by string splitting or pulling in a lib that does that? Thanks!

**Comment by RobbieMcKinstry** [comment]
Hi there @lomholdt I hope you had a lovely vacation! It's well-deserved, I'm sure!
I would personally prefer using a library. I don't have a recommendation. The top Google result [looks decent](https://www.npmjs.com/package/semver-parser).

**Comment by lomholdt** [comment]
> Hi there @lomholdt I hope you had a lovely vacation! It's well-deserved, I'm sure! I would personally prefer using a library. I don't have a recommendation. The top Google result [looks decent](https://www.npmjs.com/package/semver-parser).

Hey @RobbieMcKinstry. I noticed there's already a `semver` package installed, and I think that will do the trick. I added a warning if the parsed semver is less than `3.0.0`. You probably have an opinion about what the message should say, but I just put something in there for now. Can you give it another look? Thanks!

File: src/libs/pulumi-cli.ts [hunk 3]
```diff
@@ -42,8 +43,15 @@ export async function downloadCli(range: string): Promise<void> {
   core.info(`Configured range: ${range}`);
 
   const { version, downloads } = await getVersionObject(range);
+
   core.info(`Matched version: ${version}`);
 
+  const isUnsupportedVersion = semver.lt(version, '3.0.0');
```
Comment (cobraz) [inline src/libs/pulumi-cli.ts:49]: This looks very good. Thank you!

We should probably move `3.0.0` to a separate variable, so we can reference the same version number when we are bumping (helps with release). I can do a follow-up PR :)

**Review by cobraz — APPROVED** [review]
Thank you for this!

File: src/libs/pulumi-cli.ts (hunk 3 above)
Comment (lomholdt) [reply src/libs/pulumi-cli.ts:49]: Sound reasonable 😃 Awesome, thanks!

File: src/libs/pulumi-cli.ts [hunk 4]
```diff
@@ -42,8 +43,15 @@ export async function downloadCli(range: string): Promise<void> {
...
+
+  if (isUnsupportedVersion) {
+    core.warning(`Using Pulumi CLI version less than 3.0.0 may cause unexpected behaviour. Please use version 3.0.0 or higher.`);
```
Comment (RobbieMcKinstry) [inline src/libs/pulumi-cli.ts:52]:
```suggestion
    core.warning(`Using Pulumi CLI version less than 3.0.0 may cause unexpected behavior. Please consider migrating to 3.0.0 or higher. You can find our migration guide at https://www.pulumi.com/docs/get-started/install/migrating-3.0/`);
```

**Review by RobbieMcKinstry — APPROVED** [review]
LGTM! Thanks very much for your contribution!

**Comment by lomholdt** [comment]
> LGTM! Thanks very much for your contribution!

Awesome! Thanks 😃 No problem, and sorry for the delay.

**Comment by RobbieMcKinstry** [comment]
No need to apologize at all! I don't consider it a delay; vacations are important!

**Comment by RobbieMcKinstry** [comment]
Just FYI I don't plan to cut a release until the second week of August, which is when I would expect this change to ship. Not sure if anyone else has plans that supersede mine, though ;D

**Comment by lomholdt** [comment]
> Just FYI I don't plan to cut a release until the second week of August, which is when I would expect this change to ship. Not sure if anyone else has plans that supersede mine, though ;D

No problem. When I needed this for testing prior versions I just used my local fork and changed back to this afterwards. Looking forward to the next release 🚀
